pymc3/step_methods/sgmcmc.py

```diff
@@ -0,0 +1,273 @@
+        
+        flat_view = model.flatten(vars)
+        self.inarray = [flat_view.input]
+        self.updates = OrderedDict()
+        self.dlog_prior = prior_dlogp(vars, model, flat_view)
+        self.dlogp_elemwise = elemwise_dlogL(vars, model, flat_view)
+
+        if minibatch_tensors != None:
+            _check_minibatches(minibatch_tensors, minibatches)
+            self.minibatches = minibatches
+            
+            # Replace input shared variables with tensors
+            def is_shared(t):
+                return isinstance(t, theano.compile.sharedvalue.SharedVariable)
+            tensors = [(t.type() if is_shared(t) else t) for t in minibatch_tensors]
+            updates = OrderedDict(
+                {t: t_ for t, t_ in zip(minibatch_tensors, tensors) if is_shared(t)}
+            )
+            self.minibatch_tensors = tensors
+            self.inarray += self.minibatch_tensors
+            self.updates.update(updates)
+
+        self._initialize_values()
+        super(BaseStochasticGradient, self).__init__(vars, shared)
+
+    def _initialize_values(self):
+        """Initializes the parameters for the stochastic gradient minibatch
+        algorithm"""
+        raise NotImplementedError
+    
+    def mk_training_fn(self):
+        raise NotImplementedError
+
+    def training_complete(self):
+        """Returns boolean if astep has been called expected iter number of times"""
+        return self.expected_iter == self.t
+
+    def astep(self, q0):
+        """Perform a single update in the stochastic gradient method.
+        
+        Returns new shared values and values sampled
+        The size and ordering of q0 and q must be the same
+        Parameters
+        -------
+        q0: list
+            List of shared values and values sampled from last estimate
+
+        Returns
+        -------
+        q
+        """
+        if hasattr(self, 'minibatch_tensors'):
+            return q0 + self.training_fn(q0, *next(self.minibatches))
+        else:
+            return q0 + self.training_fn(q0)
+
+
+class SGFS(BaseStochasticGradient):
+    """
+    StochasticGradientFisherScoring
+    Implements "Algorithm 1: Bayesian Posterior Sampling via Stochastic Gradient Fisher Scoring"
+    in http://people.ee.duke.edu/%7Elcarin/782.pdf
+    """
+
+    def __init__(self, vars=None, B=None, **kwargs):
+        """
+        Parameters
+        ----------
+        vars : list 
+            Theano variables, default continuous vars
+        B : np.array
+            Symmetric positive Semi-definite Matrix
+        kwargs: passed to BaseHMC
+        """
+        self.B = B
+        super(SGFS, self).__init__(vars, **kwargs)
+
+    def _initialize_values(self):
+        # Init avg_I
+        self.avg_I = theano.shared(np.zeros((self.q_size, self.q_size)), name='avg_I')
+        self.t = theano.shared(1, name='t')
+        # 2. Set gamma
+        self.gamma = (self.batch_size + self.total_size) / (self.total_size)
```

Comment:
What is `self.input` in the code. Can you explain whats happening ? total_size is unknown till training is complete, so I am not sure how to compute it at the first sampling step.